sklearn/datasets/tests/test_base.py

```diff
@@ -381,3 +383,27 @@ def test_load_boston_alternative():
 
     np.testing.assert_allclose(data, boston_sklearn.data)
     np.testing.assert_allclose(target, boston_sklearn.target)
+
+
+def test_fetch_remote_raise_warnings_with_invalid_url():
+    if os.environ.get("SKLEARN_SKIP_NETWORK_TESTS", "1") == "1":
```

Comment:
My preference is to mock out `urlretrieve` so we do not need to touch the network. This way we can also test when the retry succeeded.

```python
def test_fetch_remote_raise_warnings_with_invalid_url(monkeypatch):
    """Check retry mechanism in _fetch_remote."""
    from urllib.error import HTTPError

    url = "https://scikit-learn.org/this_file_does_not_exist.tar.gz"
    invalid_remote_file = RemoteFileMetadata("invalid_file", url, None)
    urlretrieve_mock = Mock()
    urlretrieve_mock.side_effect = HTTPError(
        url=url, code=404, msg="Not Found", hdrs=None, fp=None
    )
    monkeypatch.setattr("sklearn.datasets._base.urlretrieve", urlretrieve_mock)

    with pytest.warns(UserWarning, match="Retry downloading") as record:
        with pytest.raises(HTTPError, match="HTTP Error 404"):
            _fetch_remote(invalid_remote_file, n_retries=3, delay=0)

        assert urlretrieve_mock.call_count == 4

        for r in record:
            assert str(r.message) == f"Retry downloading from url: {url}"
        assert len(record) == 3
```

Reply:
Good idea! Updated the code as suggested.